### DISCUSSION TITLE
Frequently asked questions about MuJoCo's renders

### DISCUSSION THREAD

**Discussion by zmsn-2077** [PKU-Alignment/omnisafe#146] [General]
Users of omnisafe, and users of MuJoCo need to visualize the trained agent or the environment, perhaps on a personal laptop, on a particular server, in a particular computing cluster, or anywhere, and we can discuss the rendering problem in different scenarios here.

**Comment by XShuheng** [comment]
Thanks for sharing excellent work on safe RL methods on the hub. But, I'm a little confused. Can this work be applied to other areas? For example, safely make the strategy to solve the problem of the power system? Like, generating safe and feasible actions that do not violate the constraints of the power system. If so, how do I change and add my environment to this code? Thanks!

**Reply by muchvo** [reply]
The answer is yes. For adding your environments to this code, you can refer to [here](https://github.com/PKU-Alignment/omnisafe/blob/main/omnisafe/envs/safety_gymnasium_env.py) to adapt your specific environments to our code without tackling more details.

**Reply by XShuheng** [reply]
Thanks for your reply. I am trying to understand this code. But, for some reason, it not works on my computer. I want to ask about the meaning of the cost after running this line code " obs, reward, cost, terminated, truncated, info = self._env.step(action.detach().cpu().numpy(),)". Is it related to the safety of the action? I don't know why this could be safe RL.

**Reply by muchvo** [reply]
Yeah, basically, the cost means in last timestep, the agent violates constraints.

**Reply by muchvo** [reply]
The main objective of this code is to enable seamless interaction between your environment and all other components within OmniSafe. Therefore, I believe that investing effort to adapt your own environment to OmniSafe is highly valuable and worthwhile.

**Reply by XShuheng** [reply]
Thanks for your response, which is very helpful for me to understand your work. By the way, how do you handle 'cost' in your code to show that this is a safeRL. Because I have adopted my environment into your code, the result still does not satisfy the safe condition. This 'cost' value is still very large.

**Reply by muchvo** [reply]
For this question, you should reference to the detail in algorithms. For example, If you use PPOLag, then you should refer to [here](https://github.com/PKU-Alignment/omnisafe/blob/main/omnisafe/algorithms/on_policy/naive_lagrange/ppo_lag.py), and you can checkout default hyperparameters [here](https://github.com/PKU-Alignment/omnisafe/blob/main/omnisafe/configs/on-policy/PPOLag.yaml).

**Reply by muchvo** [reply]
For details of how to use OmniSafe, please refer to our [tutorials](https://colab.research.google.com/github/PKU-Alignment/omnisafe/) in colab.

**Comment by XShuheng** [comment]
Thanks for your reply, and it solved my problems very well!

**Reply by muchvo** [reply]
I am happy that I can help you. If you have further questions, feel free to open a new issue or a new discussion.
